### PR TITLE
Offers

### PR DESCRIPTION
**Pull request by jackcarlisle**
# ready for review

- [x] adds basic offer flow

#36

### CODE REVIEW

File: lib/ev2_web/templates/offer/form.html.eex [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  form: f,
+  atom: :working_week,
+  options: ["5 days": 5.0, "5.5 days": 5.5, "6 days": 6.0],
+  prompt: "select..."
+  %>
+
+  <%= component "dropdown.html",
+  form: f,
+  atom: :currency,
+  options: ["GBP": "gbp", "EUR": "eur", "USD": "usd"],
```
Comment (JMurphyWeb) [inline lib/ev2_web/templates/offer/form.html.eex]: @jackcarlisle what do you think about changing these to be stored in the uppercase state? means less unnecessary formatting

File: assets/js/app.js [hunk 2]
```diff
@@ -19,3 +19,122 @@ import "phoenix_html"
...
+      if (v.target.value === "true") {
+        optionalQuestionContainer.style.display = 'none'
+      } else {
+        optionalQuestionContainer.style.display = 'block'
+      }
+    })
+  })
+}
+
+function hamburgerAnimate() {
```
Comment (JMurphyWeb) [inline assets/js/app.js]: @jackcarlisle do you mind removing all hamburger references as we don't use anymore and it's annoying that it always logs an error.

File: lib/ev2/production/offer/offer.ex [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    |> validate_length(:additional_notes, max: 300)
+    |> validate_length(:equipment_rental_description, max: 300)
+    |> validate_length(:box_rental_description, max: 300)
+  end
+
+  def validate_required_dropdowns(changeset) do
+    changeset
+    |> validate_inclusion(:daily_or_weekly, ["daily", "weekly"])
+    |> validate_inclusion(:working_week, [5.0, 5.5, 6.0])
+    |> validate_inclusion(:currency, ["gbp", "eur", "usd"])
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer.ex]: this would also need updating

File: lib/ev2/production/offer/offer.ex [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    )
+    |> validate_length(:other_deal_provisions, max: 300)
+    |> validate_length(:additional_notes, max: 300)
+    |> validate_length(:equipment_rental_description, max: 300)
+    |> validate_length(:box_rental_description, max: 300)
+  end
+
+  def validate_required_dropdowns(changeset) do
+    changeset
+    |> validate_inclusion(:daily_or_weekly, ["daily", "weekly"])
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer.ex]: We could title case this as well to reduce amount of formatting.

File: lib/ev2_web/templates/offer/form.html.eex (hunk 1 above)
Comment (jackcarlisle) [reply lib/ev2_web/templates/offer/form.html.eex]: yeah agreed!

File: assets/js/app.js (hunk 2 above)
Comment (jackcarlisle) [reply assets/js/app.js]: Yep will do

File: lib/ev2/production/offer/offer.ex (hunk 4 above)
Comment (jackcarlisle) [reply lib/ev2/production/offer/offer.ex]: sounds good

File: lib/ev2/production/offer/offer.ex (hunk 3 above)
Comment (jackcarlisle) [reply lib/ev2/production/offer/offer.ex]: cool, done

File: lib/ev2/production/calculation_helpers.ex [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      "sixth_day_fee_multiplier",
+      "seventh_day_fee_multiplier"
+    ]
+    fee_per_day_inc_holiday =
+      Map.get(offer_params, "fee_per_day_inc_holiday")
+    offer_params =
+      offer_params
+      |> update_keys(floats, &String.to_float/1)
+      |> Map.put("fee_per_day_inc_holiday", make_float(fee_per_day_inc_holiday))
+    offer_params
```
Comment (JMurphyWeb) [inline lib/ev2/production/calculation_helpers.ex]: Can we remove the offer params on this line and 129.

Comment (jackcarlisle) [reply lib/ev2/production/calculation_helpers.ex]: done

File: lib/ev2/production/offer/offer.ex [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        nil -> offer_params
+        user ->
+          Map.put_new(offer_params, "user_id", user.id)
+      end
+    project
+    |> Ecto.build_assoc(:offers)
+    |> changeset(offer_params)
+    |> Changeset.put_assoc(:contract_type, contract)
+    |> Changeset.put_assoc(:job_title, job)
+  end
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer.ex]: Not sure where this function should sit. But it has calls to Repo in it. Also would it make sense to move parts of this block of functionality into re-useable helper function and maybe higher up into a plug? Is any of it stuff we will need in all/most of the actions in offers controller?

```
+    department = Repo.get_by(Department, name: department)
 +    job = Repo.get_by(JobTitle, [name: job_title, department_id: department.id])
 +    contract_type = Production.get_contract_type(offer_params)
 +    contract = Repo.get_by(ContractType, name: contract_type)
 +    user = Repo.get_by(User, email: user_email)

```

Seems like we could extract at least the contract/job-title/department stuff into one re-useable helper function, but I'm not sure. Keen to get your thoughts.

Comment (JMurphyWeb) [reply lib/ev2/production/offer/offer.ex]: Also since `user_id` is not a required field, can we just use the `put_assoc` function to add the user to the offer?

I don't know if it works, but I was thinking remove this:
```
+    offer_params =
 +      case user do
 +        nil -> offer_params
 +        user ->
 +          Map.put_new(offer_params, "user_id", user.id)
 +      end
```

And add a line to the bottom of the changeset chain:
```
+    project
 +    |> Ecto.build_assoc(:offers)
 +    |> changeset(offer_params)
 +    |> Changeset.put_assoc(:contract_type, contract)
 +    |> Changeset.put_assoc(:job_title, job)
 +    |> Changeset.put_assoc(:users, user)
```

File: lib/ev2/production/offer/offer_api.ex [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            offer_params = Map.merge(offer_params, calculations)
+            {:ok, offer_params}
+        end
+      end
+
+      def create_offer(project, offer_params) do
+        offer_params
+        |> Offer.changeset_maybe_with_user(project)
+        |> Repo.insert()
+      end
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer_api.ex]: If we know this changeset is valid we can `insert!()` it.

File: lib/ev2/production/offer/offer_api.ex [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      ## Examples
+
+          iex> offer_form_valid?(offer_params)
+          {:ok, offer_params}
+
+      """
+      def offer_form_valid?(%{"offer" => offer_params, "project_id" => project_id}) do
+        %{"target_email" => email} = offer_params
+        offer_params =
+          Map.put(offer_params, "target_email", String.downcase(email))
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer_api.ex]: Can the three lines above be put into a changeset style formatting function like you did in the last PR?

File: lib/ev2/production/offer/offer_api.ex [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            changeset = %{changeset | action: :insert}
+            {:error, changeset}
+          true ->
+            project = Repo.get(Project, project_id)
+            calculations =
+              offer_params
+              |> CalculationHelpers.parse_offer_strings()
+              |> CalculationHelpers.run_calculations(project)
+            offer_params = Map.merge(offer_params, calculations)
+            {:ok, offer_params}
```
Comment (JMurphyWeb) [inline lib/ev2/production/offer/offer_api.ex]: I feel like this function should just return `{:ok, offer_params}`, then the controller with preloaded in project (from a plug) can run the calculations a separate function

` CalculationHelpers.run(offer_params, project)`

which just runs the following functionality:
```
+              offer_params
 +              |> CalculationHelpers.parse_offer_strings()
 +              |> CalculationHelpers.run_calculations(project)
 +            offer_params = Map.merge(offer_params, calculations)
```

File: lib/ev2/production/project/project_api.ex [hunk 10]
```diff
@@ -54,7 +54,13 @@ defmodule Ev2.Production.ProjectAPI do
           ** (Ecto.NoResultsError)
 
       """
-      def get_project!(id), do: Repo.get!(Project, id)
+      def get_project!(id) do
```
Comment (JMurphyWeb) [inline lib/ev2/production/project/project_api.ex]: Can you update documentation of this function?

File: lib/ev2_web/templates/component/offer_tile.html.eex [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    case offer.accepted do
+      nil -> "ic_pending"
+      true -> "ic_check"
+      false -> "ic_close"
+    end
+  currency =
+    case offer.currency do
+      "gbp" -> "£"
+      "eur" -> "€"
+      "usd" -> "$"
```
Comment (JMurphyWeb) [inline lib/ev2_web/templates/component/offer_tile.html.eex]: Do these need to be capitalised?

File: lib/ev2/production/offer/offer.ex (hunk 6 above)
Comment (jackcarlisle) [reply lib/ev2/production/offer/offer.ex]: Yeah we can refactor the first part you mentioned once we decide on where we can put it but let's have a discussion about it first. 

And for the second point here, there is no longer an "association" between users and offers because they exist in different `Contexts` so you have to update manually like the above.

File: lib/ev2/production/offer/offer_api.ex (hunk 8 above)
Comment (jackcarlisle) [reply lib/ev2/production/offer/offer_api.ex]: Yeah sure

File: lib/ev2/production/job_title/job_title_api.ex [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+      ## Examples
+
+          iex> list_job_titles()
+          [%JobTitles{}, ...]
+
+      """
+      def list_job_titles do
+        job_titles = Repo.all(JobTitle)
+        job_titles
+        |> Enum.map(fn job_title -> Map.from_struct(job_title) end)
```
[the remark this answers is not in the document]
Comment (jackcarlisle) [reply lib/ev2/production/job_title/job_title_api.ex]: Yeah sounds good

File: lib/ev2/production/department/department_api.ex [hunk 13]
```diff
@@ -15,8 +15,12 @@ defmodule Ev2.Production.DepartmentAPI do
           [%Departments{}, ...]
 
       """
-      def list_contract_types do
-        Repo.all(Department)
+      def list_departments do
+        departments = Repo.all(Department)
+        departments
+        |> Enum.map(fn department -> Map.from_struct(department) end)
+        |> Enum.map(fn department -> department.name end)
```
[the remark this answers is not in the document]
Comment (jackcarlisle) [reply lib/ev2/production/department/department_api.ex]: Not sure if we do actually

File: lib/ev2/production/contract_type/contract_type_helpers.ex [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+          "Supervising Prop Hand"
+        ],
+        "Stunts": [
+          "Wire Rigger"
+        ]
+    }
+
+    # if job title isn't in the list, set it to false so it will always
+    # return PAYE
+    job_titles = Map.get(replace_with_sched_d, department) || false
```
[the remark this answers is not in the document]
Comment (jackcarlisle) [reply lib/ev2/production/contract_type/contract_type_helpers.ex]: Yeah will switch it

File: lib/ev2/production/contract_type/contract_type_helpers.ex [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+      Enum.member?(project_documents, "TRANSPORT DIRECT HIRE")
+    daily_transport_paye =
+      daily && Enum.member?(project_documents, "DAILY TRANSPORT PAYE")
+    daily_transport_sch_d =
+      daily
+      && Enum.member?(project_documents, "DAILY TRANSPORT SCHEDULE-D")
+    daily_transport_direct_hire =
+      daily
+      && Enum.member?(project_documents, "DAILY TRANSPORT DIRECT HIRE")
+    daily_paye = daily && Enum.member?(project_documents, "DAILY PAYE")
```
[the remark this answers is not in the document]
Comment (jackcarlisle) [reply lib/ev2/production/contract_type/contract_type_helpers.ex]: Sounds good

File: lib/ev2/production/offer/offer_api.ex (hunk 7 above)
Comment (jackcarlisle) [reply lib/ev2/production/offer/offer_api.ex]: 👍

File: lib/ev2/production/project/project_api.ex (hunk 10 above)
Comment (jackcarlisle) [reply lib/ev2/production/project/project_api.ex]: Yep, done

File: lib/ev2_web/templates/component/offer_tile.html.eex (hunk 11 above)
Comment (jackcarlisle) [reply lib/ev2_web/templates/component/offer_tile.html.eex]: Good spot!

File: lib/ev2/production/offer/offer.ex (hunk 6 above)
Comment (JMurphyWeb) [reply lib/ev2/production/offer/offer.ex]: So perhaps that little bit of functionality can be extracted. what do you think? A function called `add_user_to_offer(user, offer)`?

File: lib/ev2_web/controllers/offer_controller.ex [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+          "new.html",
+          changeset: changeset,
+          project_id: project_id,
+          job_departments: job_departments,
+          job_title: job_title,
+          full_name: recipient_fullname
+        )
+      {:ok, offer_params} ->
+        project = Production.get_project!(project_id)
+        {:ok, offer} = Production.create_offer(project, offer_params)
```
Comment (JMurphyWeb) [inline lib/ev2_web/controllers/offer_controller.ex]: Just playing around with what I thought made sense

project = conn.assigns.project
departments = conn.assigns.departments
...

offer_params
|> Calculation_helpers.run(project)
|> add_user_if_exists()
|> Production.create_offer()
